File: frontend/src/services/postService.js
import axios from "axios";

const API_URL =
    "http://localhost:8080/posts";


// ==========================================
// GET AUTH TOKEN
// ==========================================

const getAuthHeaders = () => {

    const token =
        localStorage.getItem("token");

    if (!token) {
        return {};
    }

    return {
        Authorization:
            `Bearer ${token}`,
    };
};


// ==========================================
// POST API
// ==========================================

export const postAPI = {

    // ======================================
    // CREATE POST
    // ======================================

    createPost: async (postData) => {

        console.log(
            "POST DATA:",
            postData
        );


        const response =
            await axios.post(
                API_URL,
                postData,
                {
                    headers: {
                        ...getAuthHeaders(),

                        "Content-Type":
                            "application/json",
                    },
                }
            );


        console.log(
            "POST CREATED:",
            response.data
        );


        return response.data;
    },


    // ======================================
    // GET ALL POSTS
    // ======================================

    getAllPosts: async () => {

        const response =
            await axios.get(
                API_URL,
                {
                    headers:
                        getAuthHeaders(),
                }
            );

        return response.data;
    },


    // ======================================
    // GET POSTS BY USER
    // ======================================

    getPostsByUser: async (userId) => {

        const response =
            await axios.get(
                `${API_URL}/user/${userId}`,
                {
                    headers:
                        getAuthHeaders(),
                }
            );

        return response.data;
    },


    // ======================================
    // GET SINGLE POST
    // ======================================

    getPostById: async (id) => {

        const response =
            await axios.get(
                `${API_URL}/${id}`,
                {
                    headers:
                        getAuthHeaders(),
                }
            );

        return response.data;
    },


    // ======================================
    // UPDATE POST
    // ======================================

    updatePost: async (
        id,
        postData
    ) => {

        const response =
            await axios.put(
                `${API_URL}/${id}`,
                postData,
                {
                    headers: {
                        ...getAuthHeaders(),

                        "Content-Type":
                            "application/json",
                    },
                }
            );

        return response.data;
    },


    // ======================================
    // DELETE POST
    // ======================================

    deletePost: async (id) => {

        await axios.delete(
            `${API_URL}/${id}`,
            {
                headers:
                    getAuthHeaders(),
            }
        );
    },
};

File: backend/post-service/src/main/java/com/socialmedia/service/PostService.java
package com.socialmedia.service;

import com.socialmedia.dto.CreatePostRequest;
import com.socialmedia.dto.PostResponse;
import com.socialmedia.entity.Post;
import com.socialmedia.repository.PostRepository;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class PostService {

    private final PostRepository postRepository;


    public PostService(
            PostRepository postRepository) {

        this.postRepository =
                postRepository;
    }


    // ==========================================
    // CREATE POST
    // ==========================================

    public PostResponse createPost(
            CreatePostRequest request) {

        if (request.getUserId() == null) {

            throw new RuntimeException(
                    "User ID is required"
            );
        }


        /*
         * Text null ho to empty string
         */
        String content =
                request.getContent();

        if (content == null) {
            content = "";
        }


        /*
         * Text aur image dono empty hain
         * to post allow nahi karna
         */
        boolean hasText =
                !content.trim().isEmpty();

        boolean hasMedia =
                request.getMediaId() != null;


        if (!hasText && !hasMedia) {

            throw new RuntimeException(
                    "Post must contain text or photo"
            );
        }


        Post post =
                new Post();


        // User
        post.setUserId(
                request.getUserId()
        );


        // Content
        post.setContent(
                content.trim()
        );


        // Media
        post.setMediaId(
                request.getMediaId()
        );


        // Privacy
        if (request.getPrivacy() == null ||
                request.getPrivacy().isBlank()) {

            post.setPrivacy("Public");

        } else {

            post.setPrivacy(
                    request.getPrivacy()
            );
        }


        // Feeling
        post.setFeeling(
                request.getFeeling()
        );


        // Location
        post.setLocation(
                request.getLocation()
        );


        // Save
        Post savedPost =
                postRepository.save(post);


        return new PostResponse(
                savedPost
        );
    }


    // ==========================================
    // GET ALL POSTS
    // ==========================================

    public List<PostResponse> getAllPosts() {

        return postRepository
                .findAll()
                .stream()
                .map(PostResponse::new)
                .toList();
    }


    // ==========================================
    // GET POST BY ID
    // ==========================================

    public PostResponse getPostById(
            Long id) {

        Post post =
                postRepository
                        .findById(id)
                        .orElseThrow(() ->
                                new RuntimeException(
                                        "Post not found"
                                )
                        );

        return new PostResponse(post);
    }


    // ==========================================
    // GET POSTS BY USER
    // ==========================================

    public List<PostResponse> getPostsByUser(
            Long userId) {

        return postRepository
                .findByUserId(userId)
                .stream()
                .map(PostResponse::new)
                .toList();
    }


    // ==========================================
    // UPDATE POST
    // ==========================================

    public PostResponse updatePost(
            Long id,
            CreatePostRequest request) {

        Post post =
                postRepository
                        .findById(id)
                        .orElseThrow(() ->
                                new RuntimeException(
                                        "Post not found"
                                )
                        );


        String content =
                request.getContent();

        if (content == null) {
            content = "";
        }


        boolean hasText =
                !content.trim().isEmpty();

        boolean hasMedia =
                request.getMediaId() != null ||
                post.getMediaId() != null;


        if (!hasText && !hasMedia) {

            throw new RuntimeException(
                    "Post must contain text or photo"
            );
        }


        post.setContent(
                content.trim()
        );


        if (request.getMediaId() != null) {

            post.setMediaId(
                    request.getMediaId()
            );
        }


        if (request.getPrivacy() != null &&
                !request.getPrivacy().isBlank()) {

            post.setPrivacy(
                    request.getPrivacy()
            );
        }


        post.setFeeling(
                request.getFeeling()
        );


        post.setLocation(
                request.getLocation()
        );


        Post updatedPost =
                postRepository.save(post);


        return new PostResponse(
                updatedPost
        );
    }


    // ==========================================
    // DELETE POST
    // ==========================================

    public void deletePost(
            Long id) {

        if (!postRepository.existsById(id)) {

            throw new RuntimeException(
                    "Post not found"
            );
        }


        postRepository.deleteById(id);
    }
}